File: scripts/embedding_provider.py
#!/usr/bin/env python3
"""提供嵌入能力抽象，兼容关闭模式与 OpenAI 兼容接口。"""

from __future__ import annotations

import json
import math
from dataclasses import dataclass
from typing import Protocol
from urllib import error, request

from memory_config import EmbeddingConfig


class EmbeddingProvider(Protocol):
    """统一嵌入接口，避免写入和检索逻辑直接依赖具体服务商。"""

    @property
    def enabled(self) -> bool:
        """声明当前 provider 是否可用，方便上层快速短路。"""

    @property
    def model_name(self) -> str:
        """返回当前使用的模型名，供数据库记录和模型切换检测复用。"""

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        """按输入顺序返回向量，保证批量写入和重建阶段可稳定对齐。"""


@dataclass(frozen=True)
class DisabledEmbeddingProvider:
    """保留无嵌入模式，确保现有关键字检索流程继续工作。"""

    @property
    def enabled(self) -> bool:
        return False

    @property
    def model_name(self) -> str:
        return ""

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        return []


@dataclass(frozen=True)
class OpenAIEmbeddingProvider:
    """复用 OpenAI Embeddings API 结构，降低接入兼容服务商的成本。"""

    config: EmbeddingConfig

    @property
    def enabled(self) -> bool:
        return True

    @property
    def model_name(self) -> str:
        return self.config.model

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        """批量调用 Embeddings 接口，保证重建和查询走同一条协议链路。"""

        if not texts:
            return []
        payload = json.dumps(
            {
                "model": self.config.model,
                "input": texts,
            }
        ).encode("utf-8")
        req = request.Request(
            url=f"{self.config.base_url}/embeddings",
            data=payload,
            headers={
                "Authorization": f"Bearer {self.config.api_key}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        try:
            with request.urlopen(req, timeout=self.config.timeout_seconds) as response:
                raw = response.read().decode("utf-8")
        except error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"嵌入请求失败: HTTP {exc.code} {detail}") from exc
        except error.URLError as exc:
            raise RuntimeError(f"嵌入请求失败: {exc.reason}") from exc

        try:
            body = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise RuntimeError("嵌入请求返回了无法解析的 JSON") from exc

        data = body.get("data")
        if not isinstance(data, list):
            raise RuntimeError("嵌入响应缺少 data 字段")
        vectors: list[list[float]] = []
        for item in data:
            if not isinstance(item, dict):
                raise RuntimeError("嵌入响应数据格式不正确")
            embedding = item.get("embedding")
            if not isinstance(embedding, list):
                raise RuntimeError("嵌入响应缺少 embedding 字段")
            try:
                vectors.append([float(value) for value in embedding])
            except (TypeError, ValueError) as exc:
                raise RuntimeError("嵌入响应包含非法向量值") from exc
        if len(vectors) != len(texts):
            raise RuntimeError("嵌入响应数量与请求数量不一致")
        return vectors


def create_embedding_provider(config: EmbeddingConfig | None) -> EmbeddingProvider:
    """通过工厂函数隐藏实现细节，让调用方只依赖统一接口。"""

    if config is None:
        return DisabledEmbeddingProvider()
    return OpenAIEmbeddingProvider(config)


def build_memory_embedding_text(row) -> str:
    """拼接对检索最重要的字段，避免向量仅看到正文而忽略标题和摘要。"""

    parts = [
        f"project: {str(row['project_name'] or '').strip()}",
        f"type: {str(row['type'] or '').strip()}",
        f"title: {str(row['title'] or '').strip()}",
        f"summary: {str(row['summary'] or '').strip()}",
        f"tags: {str(row['tags'] or '').strip()}",
        str(row["content"] or "").strip(),
    ]
    return "\n".join(part for part in parts if part)


def cosine_similarity(left: list[float], right: list[float]) -> float:
    """用余弦相似度衡量语义接近度，避免向量长度差异带来偏置。"""

    if not left or not right or len(left) != len(right):
        return 0.0
    numerator = 0.0
    left_norm = 0.0
    right_norm = 0.0
    for l_value, r_value in zip(left, right):
        numerator += l_value * r_value
        left_norm += l_value * l_value
        right_norm += r_value * r_value
    if left_norm <= 0.0 or right_norm <= 0.0:
        return 0.0
    return numerator / (math.sqrt(left_norm) * math.sqrt(right_norm))

File: memory-manager/scripts/embedding_provider.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net"
	"net/http"
	"net/netip"
	"net/url"
	"regexp"
	"strings"
	"time"
)

var frontMatterRegexp = regexp.MustCompile(`\A---\n.*?\n---\n?`)

// EmbeddingProvider 抽象嵌入能力，避免搜索和写入直接耦合具体服务商。
type EmbeddingProvider interface {
	Enabled() bool
	ModelName() string
	EmbedTexts(texts []string) ([][]float64, error)
}

// DisabledEmbeddingProvider 保留无嵌入模式，确保关键字流程可独立工作。
type DisabledEmbeddingProvider struct{}

// Enabled 用显式开关让上层可以快速短路。
func (DisabledEmbeddingProvider) Enabled() bool { return false }

// ModelName 在关闭模式下返回空模型名，避免误写元数据。
func (DisabledEmbeddingProvider) ModelName() string { return "" }

// EmbedTexts 在关闭模式下不发请求，保持调用方代码统一。
func (DisabledEmbeddingProvider) EmbedTexts(texts []string) ([][]float64, error) { return nil, nil }

// OpenAIEmbeddingProvider 复用 OpenAI 兼容接口，降低接入成本。
type OpenAIEmbeddingProvider struct{ Config *EmbeddingConfig }

// Enabled 只要实例存在就视为可用，具体校验已在配置层完成。
func (p OpenAIEmbeddingProvider) Enabled() bool { return true }

// ModelName 返回当前模型名，供重建逻辑判断是否需要刷新向量。
func (p OpenAIEmbeddingProvider) ModelName() string { return p.Config.Model }

// EmbedTexts 批量调用嵌入接口，保证搜索和重建使用同一协议链路。
func (p OpenAIEmbeddingProvider) EmbedTexts(texts []string) ([][]float64, error) {
	if len(texts) == 0 {
		return nil, nil
	}
	payload, err := json.Marshal(map[string]any{"model": p.Config.Model, "input": texts})
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodPost, p.Config.BaseURL+"/embeddings", bytes.NewReader(payload))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+p.Config.APIKey)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: time.Duration(p.Config.TimeoutSeconds * float64(time.Second)), Transport: buildEmbeddingTransport(p.Config.BaseURL)}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("嵌入请求失败: %w", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("嵌入请求失败: HTTP %d %s", resp.StatusCode, strings.TrimSpace(string(body)))
	}

	var payloadBody struct {
		Data []struct {
			Embedding []float64 `json:"embedding"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &payloadBody); err != nil {
		return nil, fmt.Errorf("嵌入请求返回了无法解析的 JSON: %w", err)
	}
	if len(payloadBody.Data) != len(texts) {
		return nil, fmt.Errorf("嵌入响应数量与请求数量不一致")
	}
	vectors := make([][]float64, 0, len(payloadBody.Data))
	for _, item := range payloadBody.Data {
		if len(item.Embedding) == 0 {
			return nil, fmt.Errorf("嵌入响应缺少 embedding 字段")
		}
		vectors = append(vectors, item.Embedding)
	}
	return vectors, nil
}

// createEmbeddingProvider 通过工厂函数隐藏实现细节，让调用方只依赖统一接口。
func createEmbeddingProvider(config *EmbeddingConfig) EmbeddingProvider {
	if config == nil {
		return DisabledEmbeddingProvider{}
	}
	return OpenAIEmbeddingProvider{Config: config}
}

// buildEmbeddingTransport 私网嵌入服务优先直连，避免代理把可用服务误判成网关错误。
func buildEmbeddingTransport(baseURL string) *http.Transport {
	hostname := ""
	if parsed, err := url.Parse(baseURL); err == nil {
		hostname = parsed.Hostname()
	}
	transport := http.DefaultTransport.(*http.Transport).Clone()
	if shouldBypassProxy(hostname) {
		transport.Proxy = nil
	}
	return transport
}

// shouldBypassProxy 私网和本地地址优先直连，减少本地部署场景误走代理。
func shouldBypassProxy(hostname string) bool {
	normalized := strings.ToLower(strings.TrimSpace(hostname))
	if normalized == "" {
		return false
	}
	if normalized == "localhost" || normalized == "host.docker.internal" {
		return true
	}
	if addr, err := netip.ParseAddr(normalized); err == nil {
		return addr.IsPrivate() || addr.IsLoopback() || addr.IsLinkLocalUnicast() || addr.IsLinkLocalMulticast()
	}
	if addrs, err := net.LookupIP(normalized); err == nil {
		for _, addr := range addrs {
			if ip, ok := netip.AddrFromSlice(addr); ok && (ip.IsPrivate() || ip.IsLoopback() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast()) {
				return true
			}
		}
	}
	return false
}

// formatEmbeddingTags 把标签整理成自然文本，减少 JSON 符号带来的语义噪声。
func formatEmbeddingTags(raw any) string {
	switch value := raw.(type) {
	case []string:
		return strings.Join(value, "、")
	case string:
		text := strings.TrimSpace(value)
		if text == "" {
			return ""
		}
		if strings.HasPrefix(text, "[") && strings.HasSuffix(text, "]") {
			return strings.Join(decodeTags(text), "、")
		}
		parts := strings.Split(text, ",")
		out := make([]string, 0, len(parts))
		for _, part := range parts {
			if item := strings.TrimSpace(part); item != "" {
				out = append(out, item)
			}
		}
		return strings.Join(out, "、")
	default:
		return ""
	}
}

// stripFrontMatter 去掉持久化内容中的 YAML 头部，避免元数据重复稀释正文语义。
func stripFrontMatter(content string) string {
	text := strings.TrimSpace(content)
	if text == "" {
		return ""
	}
	return strings.TrimSpace(frontMatterRegexp.ReplaceAllString(text, ""))
}

// buildSummaryEmbeddingText 为总结记忆组织更偏结论导向的嵌入文本。
func buildSummaryEmbeddingText(row MemoryRow) string {
	parts := []string{
		"记忆类型: 总结记忆",
		"项目: " + strings.TrimSpace(row.ProjectName),
		"主题: " + strings.TrimSpace(row.Title),
		"摘要: " + strings.TrimSpace(row.Summary),
		"标签: " + formatEmbeddingTags(row.Tags),
		"正文:",
		stripFrontMatter(row.Content),
	}
	return joinNonEmpty(parts)
}

// buildErrorEmbeddingText 为错误记忆强调故障现象与修复线索，提升排障召回精度。
func buildErrorEmbeddingText(row MemoryRow) string {
	parts := []string{
		"记忆类型: 错误记忆",
		"项目: " + strings.TrimSpace(row.ProjectName),
		"问题: " + strings.TrimSpace(row.Title),
		"现象摘要: " + strings.TrimSpace(row.Summary),
		"故障标签: " + formatEmbeddingTags(row.Tags),
		"排障记录:",
		stripFrontMatter(row.Content),
	}
	return joinNonEmpty(parts)
}

// buildMemoryEmbeddingText 按记忆类型切换模板，让不同场景都保留最关键语义。
func buildMemoryEmbeddingText(row MemoryRow) string {
	if strings.EqualFold(strings.TrimSpace(row.Type), "error") {
		return buildErrorEmbeddingText(row)
	}
	return buildSummaryEmbeddingText(row)
}

// buildQueryEmbeddingText 为查询构造与目标记忆类型对齐的模板，减少短词语义损耗。
func buildQueryEmbeddingText(source string, queries []string) string {
	cleaned := make([]string, 0, len(queries))
	for _, query := range queries {
		if item := strings.TrimSpace(query); item != "" {
			cleaned = append(cleaned, item)
		}
	}
	if len(cleaned) == 0 {
		return ""
	}
	joined := strings.Join(cleaned, "、")
	if source == "error" {
		return strings.Join([]string{
			"查询类型: 错误排查记忆检索",
			"关注问题: " + joined,
			"检索目标: 查找相似的错误现象、触发条件、根因和修复结论。",
		}, "\n")
	}
	return strings.Join([]string{
		"查询类型: 总结记忆检索",
		"关注主题: " + joined,
		"检索目标: 查找相关主题、关键结论、约束条件和实现经验。",
	}, "\n")
}

// cosineSimilarity 用余弦相似度衡量语义接近度，避免向量长度差异带来偏置。
func cosineSimilarity(left, right []float64) float64 {
	if len(left) == 0 || len(left) != len(right) {
		return 0
	}
	var numerator float64
	var leftNorm float64
	var rightNorm float64
	for idx := range left {
		numerator += left[idx] * right[idx]
		leftNorm += left[idx] * left[idx]
		rightNorm += right[idx] * right[idx]
	}
	if leftNorm <= 0 || rightNorm <= 0 {
		return 0
	}
	return numerator / (math.Sqrt(leftNorm) * math.Sqrt(rightNorm))
}

// joinNonEmpty 过滤空占位文本，减少嵌入噪声。
func joinNonEmpty(parts []string) string {
	out := make([]string, 0, len(parts))
	for _, part := range parts {
		trimmed := strings.TrimSpace(part)
		if trimmed != "" && trimmed != "标签:" && trimmed != "故障标签:" {
			out = append(out, trimmed)
		}
	}
	return strings.Join(out, "\n")
}
